docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+.. _config_http_filters_lua:
+
+Lua
+===
+
+.. attention::
+
+  The Lua scripting HTTP filter is **experimental**. Use in production at your own risk. It is
+  being released for initial feedback on the exposed API and for further development, testing,
+  and verification. This warning will be removed when we feel that the filter has received enough
+  testing and API stability to call it generally production ready.
+
+Overview
+--------
+
+The HTTP Lua filter allows `Lua <https://www.lua.org/>`_ scripts to be run during both the request
+and response flows. `LuaJIT <http://luajit.org/>`_ is used as the runtime. Because of this, the
+supported Lua version is mostly 5.1 with some 5.2 features. See the `LuaJIT documentation
+<http://luajit.org/extensions.html>`_ for more details.
+
+The filter only supports loading Lua code in-line in the configuration. If local filesystem code
+is desired, a trivial in-line script can be used to load the rest of the code from the local
+environment.
+
+The design of the filter and Lua support at a high level is as follows:
+
+* All Lua environments are :ref:`per worker thread <arch_overview_threading>`. This means that
+  there is no truly global data. Any globals create and populated at load time will be visible
+  from each worker thread in isolation. True global support may be added via an API in the future.
+* All scripts are run as coroutines. This means that they are written in a synchronous style even
+  though they may perform complex asynchronous tasks. This makes the scripts substantially easier
+  to write. All network/async processing is performed by Envoy via a set of APIs. Envoy will
+  yield the script as appropriate and resume it when async tasks are complete.
+* **Do not perform blocking operations from scripts.** It is critical for performance that
+  Envoy APIs are used for all IO.
+
+Currently supported high level features
+---------------------------------------
+
+**NOTE:** It is expected that this list will expand over time as the filter is used in production.
+The API surface has been kept small on purpose. The goal is to make scripts extremely simple and
+safe to write. Very complex or high performance use cases are assumed to use the native C++ filter
+API.
+
+* Inspection of headers, body, and trailers while streaming in either the request flow, response
```

Comment:
I assume that once we have routing capabilities in filters, this is would be one of the things we'd also want to support (doesn't seem too hard, gives a lot of power..)?

Reply:
What do you mean by routing capabilities?

Reply:
@htuch do you mean giving the filter the ability to retrigger route computation after it has modified headers, etc? Almost like pre-routing in netfilter. This is turning out to be a requirement in Istio land as well, where people want to decode a JWT token, extract a user ID and then do the routing based on it. One option is to have the filter run before the base connection manager code and the route table check (but that seems complicated). Another option is to just provide an API to the the filter to kick the route computation again and store the upstream cluster in the cache (we used to have this a while ago, before we replaced it with the cached route in active stream filter).

IIUC, what @htuch is implying is that Lua filter could mutate headers (and potentially the path, doing stuff like `/foo/bar/baz -> /foo-bar-baz/api` and then recompute the routes to determine the upstream cluster. Alternatively, I can imagine doing stuff like taking a JWT token, decoding it, extracting user ID and sticking it into a Cookie header ;), and then doign header based matches.

Reply:
@rshriram Yes and yeah, we also have an internal need for this. I think you answered @mattklein123's question here, and the issue is largely orthogonal to Lua filters, but when thinking about filter APIs in general, this is something that would be good to have in Envoy, so just a friendly reminder that there is demand (and possibly cycles available to work on this at our end).

Reply:
My thinking on this is to just call `clearRouteCache()` (https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/filter.h#L118) any time a script modifies the headers to avoid scripters needing to understand the idea of route caching at all (I have to admit that when I don't need to obsess over perf at all times it's pretty nice!). I will add this in. In the future, we will likely expose the route object with a wrapper also so scripts can do more stuff based on the chosen route.